internal/cephfs/store/volumegroup.go

```diff
@@ -0,0 +1,285 @@
+	if err != nil {
+		return nil, err
+	}
+
+	if err = extractOptionalOption(&opts.NamePrefix, "volumeGroupNamePrefix", volOptions); err != nil {
+		return nil, err
+	}
+
+	opts.RequestName = req.GetName()
+
+	err = opts.Connect(cr)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err != nil {
+			opts.Destroy()
+		}
+	}()
+
+	fs := core.NewFileSystem(opts.conn)
+	opts.FscID, err = fs.GetFscID(ctx, opts.FsName)
+	if err != nil {
+		return nil, err
+	}
+
+	opts.MetadataPool, err = fs.GetMetadataPool(ctx, opts.FsName)
+	if err != nil {
+		return nil, err
+	}
+
+	return opts, nil
+}
+
+type VolumeGroupSnapshotIdentifier struct {
+	ReservedID                string
+	FsVolumeGroupSnapshotName string
+	VolumeGroupSnapshotID     string
+	RequestName               string
+	VolumeSnapshotMap         map[string]string
+}
+
+// GetVolumeIDs returns the list of volumeIDs in the VolumeSnaphotMap.
+func (vgsi *VolumeGroupSnapshotIdentifier) GetVolumeIDs() []string {
+	keys := make([]string, 0, len(vgsi.VolumeSnapshotMap))
+	for k := range vgsi.VolumeSnapshotMap {
+		keys = append(keys, k)
+	}
+
+	return keys
+}
+
+// NewVolumeGroupOptionsFromID generates a new instance of volumeGroupOptions and GroupIdentifier
+// from the provided CSI volumeGroupSnapshotID.
+func NewVolumeGroupOptionsFromID(
+	ctx context.Context,
+	volumeGroupSnapshotID string,
+	cr *util.Credentials,
+) (*VolumeGroupOptions, *VolumeGroupSnapshotIdentifier, error) {
+	var (
+		vi         util.CSIIdentifier
+		volOptions = &VolumeGroupOptions{}
+		vgs        VolumeGroupSnapshotIdentifier
+	)
+	// Decode the snapID first, to detect pre-provisioned snapshot before other errors
+	err := vi.DecomposeCSIID(volumeGroupSnapshotID)
+	if err != nil {
+		return volOptions, &vgs, cerrors.ErrInvalidVolID
+	}
+	volOptions.VolumeOptions = &VolumeOptions{}
+	volOptions.ClusterID = vi.ClusterID
+	vgs.VolumeGroupSnapshotID = volumeGroupSnapshotID
+	volOptions.FscID = vi.LocationID
+	vgs.ReservedID = vi.ObjectUUID
+
+	if volOptions.Monitors, err = util.Mons(util.CsiConfigFile, vi.ClusterID); err != nil {
+		return volOptions, &vgs, fmt.Errorf(
+			"failed to fetch monitor list using clusterID (%s): %w",
+			vi.ClusterID,
+			err)
+	}
+
+	err = volOptions.Connect(cr)
+	if err != nil {
+		return volOptions, &vgs, err
+	}
+	// in case of an error, volOptions is returned, but callers may not
+	// expect to need to call Destroy() on it. So, make sure to release any
+	// resources that may have been allocated
+	defer func() {
+		if err != nil {
+			volOptions.Destroy()
+		}
+	}()
+
+	fs := core.NewFileSystem(volOptions.conn)
+	volOptions.FsName, err = fs.GetFsName(ctx, volOptions.FscID)
+	if err != nil {
+		return volOptions, &vgs, err
+	}
+
+	volOptions.MetadataPool, err = fs.GetMetadataPool(ctx, volOptions.FsName)
+	if err != nil {
+		return volOptions, &vgs, err
+	}
+
+	err = VolumeGroupJournal.Connect(volOptions.Monitors, fsutil.RadosNamespace, cr)
```

Comment:
Is `VolumeGroupJournal` a global struct? What happens when 2 different requests are coming in at the same time, possibly even with different parameters like MONs? Can this global struct be connected multiple times without its `Destroy()` causing issues?

Reply:
Yes this can cause issues, i have updated to code to create new connections to avoid this problem